### PR TITLE
GEP-1619 Session Persistence and Session Affinity

### PR DESCRIPTION
**Pull request by gcs278**
**What type of PR is this?**
/kind gep

**What this PR does / why we need it**:
Add a GEP for session persistence that establishes the foundation for describing it and establishing a common language.

**Which issue(s) this PR fixes**:

Doesn't fix, but supports https://github.com/kubernetes-sigs/gateway-api/issues/1619

**Does this PR introduce a user-facing change?**:
N/A

Markdown Preview: https://github.com/gcs278/gateway-api/blob/gep-1619/site-src/geps/gep-1619.md

### CODE REVIEW

File: site-src/geps/gep-1619.md
Comment (howardjohn) [inline site-src/geps/gep-1619.md:33]: This seems to be saying the difference between Affinity and Persistence is L7 vs L4. Its not clear to me that is an important distinction

In my view, the primary difference between session types is whether the client _directly_ specifies the backend they want OR if some attribute of the request is just used to consistently send to a backend.

* Type 1 example: A cookie with `Backend=1.2.3.4`
* Type 2 example: Send to backend based on `backend[hash(SomeHeader) % BackendCount]`

In Envoy terms, 1 is "stick sessions" and 2 is "consistent hash".

To me, that distinction seems more relevant than L7 vs L4. I would agree that "affinity" does tend to be L4 based on source IP, as you noted in your examples.

Comment (gcs278) [reply site-src/geps/gep-1619.md]: Good point, I agree. I reworded Persistence and Affinity verbiage with focus on what information/attributes are being used to achieve the connection consistency, let me know what you think.

File: site-src/geps/gep-1619.md
Comment (sunjayBhatia) [inline site-src/geps/gep-1619.md:61]: Contour also supports using other request attributes other than cookies to route requests to a consistent backend: https://projectcontour.io/docs/v1.23.2/config/request-routing/#load-balancing-strategy (the `RequestHash` load balancing strategy)

Comment (gcs278) [reply site-src/geps/gep-1619.md]: Thanks for the info - I added a note about this in the Contour row.

File: site-src/geps/gep-1619.md
Comment (robscott) [inline site-src/geps/gep-1619.md:17]: Although this may be a non-goal for this particular "provisional" stage of the GEP, I think it would need to be a goal of the GEP as a whole. Or maybe I'm misunderstanding the purpose of writing this up.

File: site-src/geps/gep-1619.md
Comment (robscott) [inline site-src/geps/gep-1619.md:19]: If there are values that are broadly implementable, we _may_ want to consider this "core" conformance which would conflict with this non-goal. (I don't actually know that this is broadly implementable, but it may be safer to leave this non-goal out).

File: site-src/geps/gep-1619.md
Comment (robscott) [inline site-src/geps/gep-1619.md:25]: Instead of providing a history, I think I'd just explain what session affinity means and why it should be configurable within Gateway API (assuming that's the goal). Edit: Looks like you've already done that well below, so maybe just leave this part out.

File: site-src/geps/gep-1619.md
Comment (robscott) [inline site-src/geps/gep-1619.md:54]: It may help to clarify that this table intends to provide the kinds of session persistence and affinity that are supported by each implementation. Just for the sake of maintaining this going forward, it may be easiest to just focus on underlying implementations such as Envoy, NGINX, HAProxy, etc. Even if an implementation based on one of those does not support a feature today, that may just be an omission of their APIs and not a limitation of their underlying dataplane.

**Review by robscott — COMMENTED** [review]
Thanks @gcs278!

File: site-src/geps/gep-1619.md
Comment (shaneutt) [reply site-src/geps/gep-1619.md:17]: Agreed, it seems that this should be an _eventual_ goal of this KEP, however it's reasonable for the first iteration to just be the "what and the why" without the how and still merge.

**Review by shaneutt — APPROVED** [review]
Overall I really like the approach of talking about the "what" and "why" first, rather than digging right into details.

I'm approving this GEP based on the fact that it is provisional, and I see no reason to block merging it in this provisional state as I expect we can iteratively update it afterwards. That said, I do have a few comments that I would like to see resolved before we merge please see what you think of them.

File: site-src/geps/gep-1619.md
Comment (shaneutt) [reply site-src/geps/gep-1619.md:17]:
```suggestion
```

File: site-src/geps/gep-1619.md
Comment (costinm) [inline site-src/geps/gep-1619.md:31]: Privacy too. The information used for establishing the session on the internet ( the main use of the gateway ), in particular in browser context is very sensitive, used for tracking and may include user-specific info.

Comment (costinm) [reply site-src/geps/gep-1619.md:31]: I would also expand on the 'security' - maybe in separate section. Many systems use session as part of authentication - holding user info in a 'server side session' and using the session cookie as a bearer token ( Java but not only)

File: site-src/geps/gep-1619.md
Comment (costinm) [reply site-src/geps/gep-1619.md:33]: I agree, affinity is quite different - far fewer security/privacy and operational implications ( draining / lame duck / etc).

File: site-src/geps/gep-1619.md
Comment (costinm) [inline site-src/geps/gep-1619.md:47]: I'm not sure 'arbitrarily named' is right. We may support custom names, but I think Java made the right choice to define a 'default' name ( while allowing implementations to customize using their own settings). 

I would also use plural - for example in envoy-based persistence it is critical to use a second header to guarantee the same cluster ( aka Service/subset ) is used. 

I'm also not sure outside of envoy how many Internet proxies use header based persistence - would be nice having a survey. 

Using request parameter ( jsessionid=xxx in java ) is common, using the auth info is even more common. Until envoy - I haven't seen arbitrary header-based persistence.

File: site-src/geps/gep-1619.md
Comment (costinm) [inline site-src/geps/gep-1619.md:43]: In the next section you have "backend providing an arbitrary name". This section doesn't specify who provides the cookie. 

I think in both cases we need to make it a bit more clear if we want to be compatible with broadly used patterns.

1. A backend may initiate the session as needed ( Java getHttpSession(), etc) - the Gateway API should allow this and not force but allow a persistent session. In this case - the contract needs to be very clear and we should mention that the gateway implementation may alter the cookie ( to add backend info ) or have ability to decode part of the cookie (servlets and many others use a suffix to identify the node).
2. The gateway implementation itself may generate the header/cookie - in this case backend no longer has a choice to use or not sessions. 

In Java spec, a lot of discussion was around on the 'shopping cart' use case - a user may browse the store searching ( using proper load balancing ), but when he adds to cart the persistent session is created. The other important use case was auth/SSO - a page may be accessed as unauthenticated, but once user logs in usually
a cookie is used and in most implementations it creates a session. 

Having persistent sessions tied to the user is still broadly used.

File: site-src/geps/gep-1619.md
Comment (costinm) [reply site-src/geps/gep-1619.md:54]: +Java Servlets please. Persistent sessions since ~2000 encoded in the spec, multiple implementations and support in major proxies.

File: site-src/geps/gep-1619.md
Comment (costinm) [inline site-src/geps/gep-1619.md:69]: https://github.com/istio/istio/blob/master/pilot/pkg/features/pilot.go#L123

Current implementation is based on a Service label - since sessions are associated with a Service ( can't have sessions cross service due to current envoy impl ).

File: site-src/geps/gep-1619.md
Comment (costinm) [inline site-src/geps/gep-1619.md:66]: Google CloudRun - cookie based, set on service. https://cloud.google.com/run/docs/configuring/session-affinity

File: site-src/geps/gep-1619.md
Comment (pleshakov) [inline site-src/geps/gep-1619.md:73]:
```suggestion
| NGINX Kubernetes Gateway | Nginx | [Cookie-Based, Header-based  or other request attributes](https://docs.nginx.com/nginx/admin-guide/load-balancer/http-load-balancer/#enabling-session-persistence) (NGINX Plus only) |  [Hashing](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#hash) (incl. consistent hashing) based on client IP address and any other connection/request attributes. | N/A |
````

File: site-src/geps/gep-1619.md
Comment (pleshakov) [inline site-src/geps/gep-1619.md:77]: Do we need to consider how to drain established sessions?  For example, an application developer wants to upgrade backends without disrupting any established sessions but rather wait for the sessions to expire.


example of this use case -  https://github.com/jcmoraisjr/haproxy-ingress/issues/137

Comment (costinm) [reply site-src/geps/gep-1619.md:77]: Yes, Istio implementation also includes draining ( using k8s new status ) - for the rollout of a new version it is better to gradually set traffic to 0% on old version and let the connection count drop over how long you need. 

However for autoscale down - drain support and a lot of tricks are needed ( stop hook, etc). Would be great to have a common doc, I assume all vendors will have the same problems - but may be a separate discussion.

**Comment by gcs278** [comment]
Thanks for all of the feedback. I have some high priority items I'm trying to sort through at the moment, but I am planning to get back here in the next couple weeks and respond to the comments.

**Comment by gcs278** [comment]
Almost done with addressing comments. Had some urgent items come up this week. Will have a new rev posted early next week.

File: site-src/geps/gep-1619.md
Comment (gcs278) [reply site-src/geps/gep-1619.md:97]: Did you mean a new section dedicated to talking about Java? Or table entries? I started a section, but will probably need your advice on expanding on the content.

Thanks for the doc, it's very thorough. I'll leverage that.

For privacy, I think I addressed that in the new `Security and Privacy Implications` section. For compliance, could you expand? What compliance requirements are we referring to? Are you saying we should create requirements in this GEP?

**Review by gcs278 — COMMENTED** [review]
Thanks for the detailed reviews. Sorry for the delay.

**Comment by gcs278** [comment]
@shaneutt Thanks - marked back as ready for review, posted responses, and updates to the GEP.

File: site-src/geps/gep-1619.md
Comment (shaneutt) [reply site-src/geps/gep-1619.md:23]: This is sufficient, we can always improve in future PRs too if readers are asking for more clarifications :+1:

File: site-src/geps/gep-1619.md
Comment (shaneutt) [reply site-src/geps/gep-1619.md:31]: This one is kinda up to you, if to get a whole working thing you need to cover both topics, then let's do it. If not, maybe another GEP and we just make reference here? I don't consider this a blocker, so I'll consider this resolved for the purposes of this iteration, we can continue to iterate on this PR with the current lagnuage I think.

File: site-src/geps/gep-1619.md
Comment (shaneutt) [reply site-src/geps/gep-1619.md:97]: In [GEP 1709](https://gateway-api.sigs.k8s.io/geps/gep-1709/) we added an explicit [TODO Section](https://gateway-api.sigs.k8s.io/geps/gep-1709/#todo) as a list of things to follow up on in later PRs, and also as a list of things we felt would need to block moving the GEP to `implementable`. I would recommend using that same approach here to reduce how much more we need to cover in this single PR.

**Review by shaneutt — APPROVED** [review]
I think this is a great starting point, albeit we are already starting to get quite big for a first `provisional` GEP. So my suggestion is to wrap up any unresolved comments or thoughts into a TODO list (similar to https://gateway-api.sigs.k8s.io/geps/gep-1709/#todo) to mark them as things that can be picked up for follow-up PRs and that we want to resolve before `implementable`. Once that's done I think we're in a fine place to merge this as `Provisional` and start thinking about the next iterative steps.

**Review by sunjayBhatia — APPROVED** [review]
lgtm given the stated goals

File: site-src/geps/gep-1619.md
Comment (gcs278) [reply site-src/geps/gep-1619.md:31]: Sounds good. I'm going to say lets keep it as is, and I think we can always diverge into another GEP in the future if it becomes too disconnected.

File: site-src/geps/gep-1619.md
Comment (gcs278) [reply site-src/geps/gep-1619.md:97]: Adding to a TODO section.

**Comment by gcs278** [comment]
I've created a `TODO` section and added TODOs from the comments from the review, mostly from @costinm's comments. Let me know if this looks good.

**Review by shaneutt — APPROVED** [review]
With the TODO list in place for things we need to resolve before `implementable`, the time this has been up for review, the fact that it's provisional, and that nobody has any blocking comments it seems like now is a good time to get this first pass landed and we can build off from here.

Thanks @gcs278!

/lgtm

**Pull request merged**
